tools/degreeworks-scraper/src/components/Scraper.ts

```diff
@@ -0,0 +1,190 @@
+    for (const [, { specs, school, code: majorCode, degreeType: degree }] of [
+      ...this.parsedUgradPrograms,
+      ...this.parsedGradPrograms,
+    ]) {
+      if (!degree) throw new Error("Degree type is undefined");
+      for (const specCode of specs) {
+        const audit = await this.dw.getSpecAudit(degree, school, majorCode, specCode);
+        if (!audit) {
+          console.log(
+            `Requirements block not found (school = ${school}, majorCode = ${majorCode}, specCode = ${specCode}, degree = ${degree})`,
+          );
+          continue;
+        }
+        this.parsedSpecializations.set(
+          specCode,
+          this.ap.parseBlock(`${school}-SPEC-${specCode}-${degree}`, audit),
+        );
+        console.log(
+          `Requirements block found and parsed for "${audit.title}" (specCode = ${specCode})`,
+        );
+      }
+    }
+    this.degreesAwarded = new Map(
+      Array.from(
+        new Set(
+          [...this.parsedUgradPrograms, ...this.parsedGradPrograms].map(([, x]) => x.degreeType!),
+        ),
+      ).map((x) => [x, this.degrees!.get(x)!]),
+    );
+
+    // Post-processing steps.
+
+    // As of this commit, the only program which seems to require both of
+    // its "specializations" is the B.A. in Art History. There's probably a
+    // cleaner way to address this, but this is such an insanely niche case
+    // that it's probably not worth the effort to write a general solution.
+
+    let x, y, z;
+    if (
+      (x = this.parsedUgradPrograms.get("Major in Art History")!) &&
+      (y = this.parsedSpecializations.get("AHGEO")!) &&
+      (z = this.parsedSpecializations.get("AHPER")!)
+    ) {
+      x.specs = [];
+      x.requirements = { ...x.requirements, ...y.requirements, ...z.requirements };
+      this.parsedSpecializations.delete("AHGEO");
+      this.parsedSpecializations.delete("AHPER");
+      this.parsedUgradPrograms.set("Major in Art History", x);
+    }
+
+    // Some programs have an empty requirements block and more than one specialization.
+    // They can be simplified into a "Select 1 of the following" group requirement.
+    this.parsedUgradPrograms = this.cleanUpPrograms(this.parsedUgradPrograms);
+    this.parsedGradPrograms = this.cleanUpPrograms(this.parsedGradPrograms);
+
+    this.done = true;
+  }
+  get() {
+    if (!this.done) throw new Error("This scraper instance has not yet finished its run.");
+    return {
+      parsedMinorPrograms: this.parsedMinorPrograms!,
+      parsedUgradPrograms: this.parsedUgradPrograms!,
+      parsedGradPrograms: this.parsedGradPrograms!,
+      parsedSpecializations: this.parsedSpecializations!,
+      degreesAwarded: this.degreesAwarded!,
+    };
+  }
+  static async new(authCookie: string): Promise<Scraper> {
+    const studentId = jwtDecode<JwtPayload>(authCookie.slice("Bearer+".length))?.sub;
+    if (!studentId || studentId.length !== 8)
+      throw new Error("Could not parse student ID from auth cookie.");
```

Comment:
I think the null case is invalid anyways, so just an optional chain comparison directly to 8 should work?

```suggestion
    if (studentId?.length !== 8)
      throw new Error("Could not parse student ID from auth cookie.");
```